server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/DeletedMessageVaultRestoreTask.java

```diff
@@ -0,0 +1,174 @@
+        private final long errorRestoreCount;
+        private final long failRemovingCount;
+
+        AdditionalInformation(User user, long successfulRestoreCount, long errorRestoreCount, long failRemovingCount) {
+            this.user = user;
+            this.successfulRestoreCount = successfulRestoreCount;
+            this.errorRestoreCount = errorRestoreCount;
+            this.failRemovingCount = failRemovingCount;
+        }
+
+        public long getSuccessfulRestoreCount() {
+            return successfulRestoreCount;
+        }
+
+        public long getErrorRestoreCount() {
+            return errorRestoreCount;
+        }
+
+        public long getFailRemovingCount() {
+            return failRemovingCount;
+        }
+
+        public String getUser() {
+            return user.asString();
+        }
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DeletedMessageVaultRestoreTask.class);
+
+    static final String TYPE = "deletedMessages/restore";
+    static final String RESTORE_MAILBOX_NAME = "restored messages";
+
+    private final DeletedMessageVault deletedMessageVault;
+    private final User userToRestore;
+    private final MailboxManager mailboxManager;
+    private final MailboxPath restoreMailbox;
+
+    private final AtomicLong successfulRestoreCount;
+    private final AtomicLong errorRestoreCount;
+    private final AtomicLong failRemovingCount;
+
+    DeletedMessageVaultRestoreTask(DeletedMessageVault deletedMessageVault, User userToRestore, MailboxManager mailboxManager) {
+        this.deletedMessageVault = deletedMessageVault;
+        this.userToRestore = userToRestore;
+        this.mailboxManager = mailboxManager;
+        this.restoreMailbox = MailboxPath.forUser(userToRestore.asString(), RESTORE_MAILBOX_NAME);
+        this.successfulRestoreCount = new AtomicLong();
+        this.errorRestoreCount = new AtomicLong();
+        this.failRemovingCount = new AtomicLong();
+    }
+
+    @Override
+    public Result run() {
+        try {
+            MailboxSession session = mailboxManager.createSystemSession(userToRestore.asString());
+            MessageManager restoreMailboxManager = restoreMailboxManager(session);
+
+            return Flux.from(deletedMessageVault.search(userToRestore, Query.ALL)).toStream()
+                .map(deletedMessage -> moveToMailbox(restoreMailboxManager, deletedMessage, session))
+                .reduce(Result.COMPLETED, Task::combine);
+        } catch (MailboxException e) {
+            LOGGER.error("Error happens when while creating messageManager for mailbox {}", restoreMailbox.asString(), e);
+            return Result.PARTIAL;
+        }
+    }
+
+    private Result moveToMailbox(MessageManager restoreMailboxManager, DeletedMessage deletedMessage, MailboxSession session) {
+        Result appended = appendToMailbox(restoreMailboxManager, deletedMessage, session);
+
+        return Optional.of(appended)
```

Comment:
what if you implement `map` in Result instead?

Reply:
removed Task Result combining stuffs